### PR TITLE
Add a uid/gid flag to spawn a container with non-zero uid/gid.

### PR DESCRIPTION
**Pull request by maleadt**
Turns out `tar` behaves differently when `id==0` -- https://github.com/habitat-sh/builder/issues/365#issuecomment-382862233, which doesn't work in the single-user containers we spawn (breaking BinaryProvider) -- so add an option to create a user map that uses a different uid/gid.

Fixes https://github.com/staticfloat/Sandbox.jl/issues/24

### CODE REVIEW

**Comment by maleadt** [comment]
```
/home/runner/.julia/artifacts/b75badd28f02d9a4824aa829a7537d1c6f39fb13/bin/sandbox: unrecognized option '--uid'
```

Doesn't look like the `build local sandbox` CI is actually using the local sandbox. Presumable `Pkg.test` uses a different environment that ignores the `LocalPreferences.toml`?

**Comment by giordano** [comment]
> Presumable Pkg.test uses a different environment that ignores the LocalPreferences.toml?

Yes: https://github.com/JuliaLang/Pkg.jl/issues/2500.  You should have a separate [`test/Project.toml`](https://github.com/JuliaMath/FFTW.jl/blob/53fe68911045932f56b77c38c068eab64617f690/test/Project.toml) like `FFTW.jl`

File: test/Project.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+[deps]
+Artifacts = "56f22d72-fd6d-98f1-02f0-08ddc0907c33"
+JLLWrappers = "692b3bcd-3c85-4b1f-b108-f13ce0eb3210"
+Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+Preferences = "21216c6a-2e73-6563-6e65-726566657250"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Scratch = "6c6a2e73-6563-6170-7368-637461726353"
+TOML = "fa267f1f-6049-4f14-aa54-33bafae1ed76"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+UserNSSandbox_jll = "b88861f7-1d72-59dd-91e7-a8cc876a4984"
```
Comment (giordano) [inline test/Project.toml:14]: If I understand test-specific projects correctly, you only need to specify the packages you actually use, excluding the current package itself (this is at least what we did in `FFTW.jl`)
```suggestion
Artifacts = "56f22d72-fd6d-98f1-02f0-08ddc0907c33"
Scratch = "6c6a2e73-6563-6170-7368-637461726353"
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```

**Comment by maleadt** [comment]
Doesn't seem to help here.

**Comment by giordano** [comment]
Maybe [this](https://github.com/JuliaMath/FFTW.jl/blob/53fe68911045932f56b77c38c068eab64617f690/.github/workflows/CI.yml#L63-L64) is relevant?

**Comment by maleadt** [comment]
Ah, yes, so that's setting a global preferences. Seems pretty hacky, at this point we could just run `julia --project test/runtests.jl` and just not use `Pkg.test`...

**Comment by maleadt** [comment]
OK, only legit failures now.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-using Test, Sandbox
+using Test, Sandbox, UserNSSandbox_jll
```
Comment (maleadt) [inline test/runtests.jl:1]: It's annoying how a project needs to depend on a package for Preferences to pick up the preference. Is that intended?

**Comment by maleadt** [comment]
OK, all done. How do we want to proceed here? Merge with failures, rebuild UserNSSandbox_jll, and re-run CI to verify?

**Comment by maleadt** [comment]
@staticfloat, re. https://github.com/staticfloat/Sandbox.jl/issues/24#issuecomment-822747930, I take it you're opposed to this? I don't see an alternative though, `su` or similar doesn't work in the userns, and `id=0` really does make software behave differently.

**Review by staticfloat — CHANGES_REQUESTED** [review]

**Comment by staticfloat** [comment]
> I take it you're opposed to this?

Actually, I didn't really think that the patch to `userns_sandbox.c` would be so small, so I'm totally open to this change.  In fact, it may be quite useful for BB to build things as non-root!

**Comment by giordano** [comment]
>  it may be quite useful for BB to build things as non-root!

I thought having a single user helped with mapping permissions :thinking:

**Comment by staticfloat** [comment]
Indeed it does, but this nicely dodges the problem by keeping things as a single-user system, it just allows using a UID that is nonzero.  I think we will want to have the BB sandbox default to using the same numeric ID as the originating user, which will even allow us to avoid the `chown` calls we have to make after the Docker runner finishes.

File: src/Docker.jl [hunk 3]
```diff
@@ -91,7 +91,10 @@ function build_docker_image(root_path::String; verbose::Bool = false)
 
         # Build the docker image
         open(`docker import - $(image_name)`, "w", verbose ? stdout : devnull) do io
-            Tar.create(root_path, io)
+            # we need to record permisions, so can't use Tar.jl
+            cd(root_path) do
+                run(pipeline(`tar -c .`, stdout=io))
+            end
```
[the remark this answers is not in the document]
Comment (maleadt) [reply src/Docker.jl:97]: There's details in the commit message. I've created an issue too: https://github.com/JuliaIO/Tar.jl/issues/103

File: src/Docker.jl [hunk 4]
```diff
@@ -163,6 +166,11 @@ function build_executor_command(exe::DockerExecutor, config::SandboxConfig, user
         append!(cmd_string, ["--entrypoint", config.entrypoint])
     end
 
+    # Set the user and group, if requested
+    if config.uid !== nothing || config.gid !== nothing
+        append!(cmd_string, ["--user", "$(something(config.uid, 0)):$(something(config.gid, 0))"])
```
[the remark this answers is not in the document]
Comment (maleadt) [reply src/Docker.jl:171]: Actually, turns out that Docker does support a group-only specification, so we could keep both fields optional. Or have them default to 0 for consistency, same to me.

File: test/runtests.jl (hunk 2 above)
Comment (maleadt) [reply test/runtests.jl]: Yeah, I (mistakenly) thought you needed the import too.

**Comment by staticfloat** [comment]
I'm quite happy with this, with the one exception that using `Tar.jl` was one part of the hope that I had that this package would be able to work on Windows without further binary dependencies. :(

I guess we can work towards that in the future, in the `Tar.jl` repo.

**Pull request merged**